src/analysis/processing/qgsalgorithmextractlabels.cpp

```diff
@@ -0,0 +1,372 @@
+    QgsProcessingFeedback *mFeedback = nullptr;
+    QList<QgsFeature> features;
+};
+
+QVariantMap QgsExtractLabelsAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  // MAP_THEME has been taken care of in prepareAlgorithm
+  QgsCoordinateReferenceSystem crs = parameterAsExtentCrs( parameters, QStringLiteral( "EXTENT" ), context );
+  if ( !crs.isValid() )
+    crs = context.project()->crs();
+  const QgsRectangle extent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context );
+  const double scale = parameterAsDouble( parameters, QStringLiteral( "SCALE" ), context );
+  const double dpi = parameterAsDouble( parameters, QStringLiteral( "DPI" ), context );
+
+  QgsScaleCalculator calculator;
+  calculator.setDpi( dpi );
+  calculator.setMapUnits( crs.mapUnits() );
+  const QSizeF imageSize = calculator.calculateImageSize( extent, scale );
+
+  QgsFields fields;
+  fields.append( QgsField( QStringLiteral( "layer" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "featureid" ), QVariant::LongLong, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "label_text" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "label_alignment" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "label_rotation" ), QVariant::Double, QString(), 20, 10 ) );
+  fields.append( QgsField( QStringLiteral( "font_family" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "font_style" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "font_size" ), QVariant::Double, QString(), 20, 10 ) );
+  fields.append( QgsField( QStringLiteral( "font_italic" ), QVariant::Bool, QString(), 1, 0 ) );
+  fields.append( QgsField( QStringLiteral( "font_bold" ), QVariant::Bool, QString(), 1, 0 ) );
+  fields.append( QgsField( QStringLiteral( "font_letter_spacing" ), QVariant::Double, QString(), 20, 10 ) );
+  fields.append( QgsField( QStringLiteral( "font_word_spacing" ), QVariant::Double, QString(), 20, 10 ) );
+  fields.append( QgsField( QStringLiteral( "format_line_height" ), QVariant::Double, QString(), 20, 10 ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, QgsWkbTypes::Point, crs, QgsFeatureSink::RegeneratePrimaryKey ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  QgsMapSettings mapSettings;
+  mapSettings.setDestinationCrs( crs );
+  mapSettings.setExtent( extent );
+  mapSettings.setOutputSize( imageSize.toSize() );
+  mapSettings.setOutputDpi( dpi );
+  mapSettings.setFlag( Qgis::MapSettingsFlag::DrawLabeling, true );
+  mapSettings.setFlag( Qgis::MapSettingsFlag::SkipSymbolRendering, true );
```

Comment:
I think you should also set the UseRenderingOptimization flag here -- that will speed up the algorithm a LOT

Reply:
Great piece of wisdom :)